File: library/src/main/java/me/shkschneider/skeleton/network/GZipRequest.java
package me.shkschneider.skeleton.network;

import android.support.annotation.NonNull;

import com.android.volley.NetworkResponse;
import com.android.volley.ParseError;
import com.android.volley.Response;
import com.android.volley.toolbox.HttpHeaderParser;
import com.android.volley.toolbox.StringRequest;

import java.io.BufferedReader;
import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.zip.GZIPInputStream;

// <https://github.com/DWorkS/VolleyPlus>
public class GZipRequest extends StringRequest {

    public GZipRequest(final int method, @NonNull final String url, final Response.Listener<String> listener, final Response.ErrorListener errorListener) {
        super(method, url, listener, errorListener);
    }

    public GZipRequest(@NonNull final String url, final Response.Listener<String> listener, final Response.ErrorListener errorListener) {
        super(url, listener, errorListener);
    }

    @Override
    protected Response<String> parseNetworkResponse(@NonNull final NetworkResponse response) {
        final StringBuilder stringBuilder = new StringBuilder();
        try {
            final GZIPInputStream gzipInputStream = new GZIPInputStream(new ByteArrayInputStream(response.data));
            final InputStreamReader inputStreamReader = new InputStreamReader(gzipInputStream);
            final BufferedReader bufferedReader = new BufferedReader(inputStreamReader);
            String line;
            while ((line = bufferedReader.readLine()) != null) {
                stringBuilder.append(line);
            }
            inputStreamReader.close();
            bufferedReader.close();
            gzipInputStream.close();
        }
        catch (final IOException e) {
            return Response.error(new ParseError(e));
        }
        return Response.success(stringBuilder.toString(), HttpHeaderParser.parseCacheHeaders(response));
    }

}

File: library/src/main/kotlin/me/shkschneider/skeleton/network/GZipRequest.kt
package me.shkschneider.skeleton.network

import com.android.volley.NetworkResponse
import com.android.volley.ParseError
import com.android.volley.Response
import com.android.volley.toolbox.HttpHeaderParser
import com.android.volley.toolbox.StringRequest

import java.io.BufferedReader
import java.io.ByteArrayInputStream
import java.io.IOException
import java.util.zip.GZIPInputStream

// <https://github.com/DWorkS/VolleyPlus>
class GZipRequest : StringRequest {

    constructor(method: Int, url: String, listener: Response.Listener<String>, errorListener: Response.ErrorListener) : super(method, url, listener, errorListener)

    constructor(url: String, listener: Response.Listener<String>, errorListener: Response.ErrorListener) : super(url, listener, errorListener)

    override fun parseNetworkResponse(response: NetworkResponse): Response<String> {
        try {
            val gzipInputStream = GZIPInputStream(ByteArrayInputStream(response.data))
            // <https://stackoverflow.com/a/41000650>
            val string = gzipInputStream.bufferedReader().use(BufferedReader::readText)
            gzipInputStream.close()
            return Response.success(string, HttpHeaderParser.parseCacheHeaders(response))
        } catch (e: IOException) {
            return Response.error(ParseError(e))
        }
    }

}
